packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -3,438 +3,363 @@
+    console.log(`currentRightAllocation`, tryRight.seat.getCurrentAllocation());
+    console.log(`proposed left reallocation`, leftAllocation);
+    console.log(`proposed right reallocation`, rightAllocation);
+    // show the constraints
+    console.log(`left want`, keepLeft.seat.getProposal().want);
+    console.log(`right want`, tryRight.seat.getProposal().want);
+
+    if (!offerSafeForLeft) {
+      console.log(`offer not safe for left`);
+    }
+    if (!offerSafeForRight) {
+      console.log(`offer not safe for right`);
+    }
+    return tryRight.seat.kickOut();
+  }
 
-      // Check whether reallocate would error before calling. If
-      // it would error, reject the right offer and return.
-      const offerSafeForLeft = helpers.isOfferSafe(
-        keepLeft.offerHandle,
-        leftAllocation,
-      );
-      const offerSafeForRight = helpers.isOfferSafe(
-        tryRight.offerHandle,
-        rightAllocation,
-      );
-      if (!(offerSafeForLeft && offerSafeForRight)) {
-        console.log(
-          `currentLeftAllocation`,
-          zcf.getCurrentAllocation(keepLeft.offerHandle),
-        );
-        console.log(
-          `currentRightAllocation`,
-          zcf.getCurrentAllocation(tryRight.offerHandle),
-        );
-        console.log(`proposed left reallocation`, leftAllocation);
-        console.log(`proposed right reallocation`, rightAllocation);
-        // show the contraints
-        console.log(
-          `left want`,
-          zcf.getOffer(keepLeft.offerHandle).proposal.want,
-        );
-        console.log(
-          `right want`,
-          zcf.getOffer(tryRight.offerHandle).proposal.want,
-        );
+  return zcf.reallocate(
+    keepLeft.seat.stage(leftAllocation),
+    tryRight.seat.stage(rightAllocation),
+  );
+};
 
-        if (!offerSafeForLeft) {
-          console.log(`offer not safe for left`);
-        }
-        if (!offerSafeForRight) {
-          console.log(`offer not safe for right`);
-        }
-        return rejectOffer(tryRight.offerHandle);
-      }
-      zcf.reallocate(
-        [keepLeft.offerHandle, tryRight.offerHandle],
-        [leftAllocation, rightAllocation],
-      );
-      return undefined;
+/**
+ * If the two handles can trade, then swap their compatible assets,
+ * marking both offers as complete.
+ *
+ * The surplus remains with the original offer. For example if
+ * offer A gives 5 moola and offer B only wants 3 moola, offer A
+ * retains 2 moola.
+ *
+ * If the keep offer is no longer active (it was already completed), the try
+ * offer will be rejected with a message (provided by 'keepHandleInactiveMsg').
+ *
+ * TODO: If the try offer is no longer active, swap() should terminate with
+ * a useful error message, like defaultRejectMsg.
+ *
+ * If the swap fails, no assets are transferred, and the 'try' offer is rejected.
+ *
+ * @param {ZCFSeat} keepSeat
+ * @param {ZCFSeat} trySeat
+ * @param {String} [keepHandleInactiveMsg]
+ */
+export const swap = (
```

Comment:
I wonder if we really need `trade` and `swap` as distinct abstractions. Swap begins by checking if keepSeat has exited. However, it should also error if trySeat has exited. `trade` should also error if either has exited, so both checks could be moved into `trade`.

Looking at the callers of `trade`, a major reason they can't use `swap` is that the `keepSeat` remains open. They all do exit the trySeat. 

Looking at the callers of `swap`, since `swap` throws on failure and only closes the `keepSeat` on success, it would not be an undue burden for them to exit the keepSeat after the call to swap. Since we did presciently call it the `keepSeat`, it would also not be surprising for `swap` itself not to close it.